src/OrleansRuntime/Placement/IActivationSelector.cs

```diff
@@ -9,10 +10,13 @@ internal interface IActivationSelector
     {
         Task<PlacementResult> OnSelectActivation(
             PlacementStrategy strategy, GrainId target, IPlacementRuntime context);
+
+        bool TrySelectActivationSynchronously(
```

Comment:
It'll require adding System.Threading.Tasks.Extensions package, which most likely would be out of scope of this PR. Plus, the problem is not the return type of ```Task``` but ```await```s in method body which causes allocation of state machine with its costs.